### ISSUE TITLE
Large target_exe files cause timeouts when creating libfuzzer jobs

### ISSUE THREAD

**Issue by puhley** [microsoft/onefuzz#1477]
## Information

- Onefuzz version: 3.2
- OS: Mac/Linux

## Provide detailed reproduction steps (if any)

1. When starting a libfuzzer job with a large target_exe (around 300MB) via _onefuzz template libfuzzer basic..._, a series of timeout errors occur:
```
INFO:onefuzz:uploading target exe \`./my_large_target.exe\`
WARNING:nsv-backend:upload_file failed with ServiceResponseError("('Connection aborted.', timeout('The write operation timed out'))"), retrying ...
WARNING:nsv-backend:upload_file failed with ServiceResponseError("('Connection aborted.', timeout('The write operation timed out'))"), retrying ...
...
```
2. The nsv-backend error repeats until it hits the max retries, resulting in a final error:
```
...
WARNING:nsv-backend:upload_file failed with ServiceResponseError("('Connection aborted.', timeout('The write operation timed out'))"), retrying ...
ERROR:cli:command failed: ('Connection aborted.', timeout('The write operation timed out'))
```

### Expected result

Creating a OneFuzz job with a large target_exe should succeed. 

### Actual result

The upload of a large target_exe results in a timeout.

### Proposed solutions
In testing, we've noticed that lowering the _max_concurrency_ fixes the issue.  The _max_concurrency_ is defined within the OneFuzz code here: https://github.com/microsoft/onefuzz/blob/7c86a3d7d91f07c95b05db1f4181618dab2c7cbd/src/cli/onefuzz/backend.py. It is a parameter that is provided to the Azure Python SDK _upload_blob_ method:  https://docs.microsoft.com/en-us/python/api/azure-storage-blob/azure.storage.blob.containerclient?view=azure-python#upload-blob-name--data--blob-type--blobtype-blockblob---blockblob----length-none--metadata-none----kwargs- .  According to the documentation, it specifies the, "Maximum number of parallel connections to use when the blob size exceeds 64MB."  This would apply to our scenario since the target_exe is larger than 64MB.

I believe what is occurring is that Azure Python SDK is creating multiple async upload streams. For a low bandwidth connection, these streams exceed the maximum upload bandwidth. Imagine a freeway where multiple parallel lanes of traffic get restricted to a single lane of traffic causing everyone to slow down. Data Stream 1 has to keep pausing its work to let packets from Data Streams 2 -> X also go through. This causes Data Stream 1 to take longer than normal to complete which results in a timeout. If you reduce the max_concurrency, then Data Stream 1 can finish in its expected amount of time because it doesn't have to share the road with other concurrent data streams. While we have had issues with _max_concurrency_ in our environments, I am assuming that there are environments where it is not an issue.

The Azure SDK _upload_blob_ function also has a timeout parameter that is expressed in seconds. In my testing, I tried setting it as high as 500, but it did not appear to have any affect on the problem.

One way to address this would be to make max_concurrency a configurable parameter. It could still default to 10 for organizations that have high bandwidth connections. However, people working from home or other restricted bandwidth environments could tune max_concurrency down to their bandwidth capacity. Although, this may result in excessive command line parameters and may require too much internal knowledge of OneFuzz to understand.

A different approach would be to have the function detect when a retry event is thrown and reduce the max_concurrency with each retry attempt. The max_concurrency defaults to 10 and there are 10 retry attempts. Therefore, it could walk its way down to a max_concurrency of 1. This has the downside of the user having to wait longer for the command to complete as it tries to find a suitable max_concurrency.  However, the command would eventually work on the first try.

A third-option would be to make it an advanced setting that you could specify using _onefuzz config_. Affected environments could set it once and leave it set. I am not sure if that is the type of configuration setting that the config functionality is intended to manage.

If an approach can be approved, then I could write a pull request for review.
 

AB#39993988

**Comment by ranweiler** [comment]
Thanks for the report, @puhley.

I'm able to minimize to the following repro steps:
```bash
# Create a big blob.
dd if=/dev/zero of=zeros.bin bs=300M count=1

# Create destination container.
onefuzz containers create big

# Simulate a low-bandwidth link.
sudo tc qdisc add dev eth0 root tbf rate 104kbit latency 50ms burst 1540

# Try to upload blob.
onefuzz containers files upload_file big -v zeros.bin
```
If you are able to share any info about the CPU count and upload bandwidth on the hosts where you experience this, it would be helpful in scoping this issue.

My questions:
1. Does this represent a bug in the underlying Azure SDK for Python, which needs reporting?
2. Even if it is technically a bug, what is the right workaround?

Re (1), the name `max_concurrency` make me wonder if the burden should be on the blob storage client library to scale up conservatively, up to (but not exceeding) `max_concurrency`. In practice, it seems to be used to exactly _set_ the number of thread pool workers in the uploader. As a corollary: if we're setting the value to 10, but the default is the host CPU count, perhaps we should NOT set it, and let high-bandwidth/CPU users crank it up.

Possibly-related (now-closed) upstream issues:
- https://github.com/Azure/azure-sdk-for-python/issues/20447
- https://github.com/Azure/azure-sdk-for-python/issues/10572

Re (2), I think a tunable as a CLI arg is probably the best, least-intrusive workaround. @chkeita, WDYT?

**Comment by puhley** [comment]
I don't have stats for all of the places that have encountered this issues. One of the locations has 5Mbps upload speed from a MacBook Pro 2019 (2.3 GHz 8-Core Intel Core i9, 16GB ram). I may have also reproduced in a 10Mbps upload environment but I am not sure since I didn't measure at the time. 

In looking at the Azure SDK on GitHub, max_concurrency appears to have a default of 'None' or '1' in most of the search results which suggests that single thread uploads are the default.  I am assuming that OneFuzz's default of 10 for concurrency works if you are transferring large binaries between different Azure services where bandwidth speeds are higher. For instance, the Standard_D2ds_v4 size that OneFuzz uses in Azure has 1000Mbps throughput. For people working from home, a lower tier Internet plan will have an upload rate of somewhere between 4-10Mbps upload speeds depending on your provider. You could also be restricted if you are using a corporate VPN at the time.

I don't know the Azure SDK well enough to know which source file is definitive. I am basing my assumptions off of the following code and the subsequent files it calls:
https://github.com/Azure/azure-sdk-for-python/blob/0c016cf35670aa0f58611e519442fae63d77f0ab/sdk/storage/azure-storage-blob/azure/storage/blob/aio/_blob_client_async.py#L46

The code appears to be fairly simple in establishing the concurrent threads. If you ask for 10 parallel threads, then it uses Python's concurrent.futures to establish 10 threads. It doesn't try to add any monitoring logic on top of that.

In doing some back-of-the-napkin arithmetic, the default chunk size of 4MB via a 5Mbps upload connection would take roughly 6.4 seconds in ideal conditions. If it is sharing the road with 9 other streams, it would take roughly 64 seconds to upload 40MBs across the 10 streams and a 300MB upload would take 7 and a half minutes in ideal conditions (ignoring other processes on the device, shared network, etc.). I tried looking up Azure Blob timeouts here: https://docs.microsoft.com/en-us/rest/api/storageservices/setting-timeouts-for-blob-service-operations. Since the file is being transferred in chunks, the timeouts would apply to each chunk. However, I don't know the Blob service well enough to know which of them would be triggering the timeout in this scenario.

**Comment by ranweiler** [comment]
To recap, I think we want a solution that:
1. Succeeds (if slowly) on reasonably low-bandwidth connections, like in your scenario.
2. Is as fast as possible when bandwidth is high.
3. Avoids leaking implementation details in the CLI API

I did some additional experiments, and configured my traffic shaping to repro both your failure and your success case (when `max_connections` is hardcoded to 1). This was tested on an Ubuntu 18.04 VM with 8 vCPUs and 16 GB of RAM (similar in spec to your MacBook scenario, just fewer logical cores).

I shaped traffic via:
```
sudo tc qdisc add dev eth0 root tbf rate 1mbps latency 50ms burst 1540
```

In this situation (as you report):
1. `max_concurrency=1` succeeds (around 5m50s of wall clock time)
2. `max_concurrency=10` fails, with or without a `timeout` of 500 seconds
3. _Removing_ the `max_concurrency` arg succeeds (~5m50s)
4. Using `azcopy cp` with a SAS URL succeeds in 5m45s (AzCopy 10.13)

**Note:** we already use (and require) `azcopy` for container syncing, both in templates and in direct CLI calls to `files upload_dir`.

When I disable traffic shaping:
1. `max_concurrency=1` succeeds (~50s)
2.  `max_concurrency=10` succeeds (8s)
3. No `max_concurrency` succeeds (~50s)
4. `azcopy cp` succeeds (4s)

### Conclusion

I think we should:
1. Update our use of `max_connections` to be more conservative (probably: remove it)
2. Use `azcopy` whenever available, in `upload_file`.

`azcopy` gets us the fastest uploads in both low and high bandwidth scenarios, and our existing code can be reused for this single-blob case. Our wrapper code also currently uses `sync`, so it will avoid redundant reuploads (which we _don't_ avoid right now).

We might even prefer to eventually _require_ `azcopy`, but that seems unnecessary at the moment. I do like the idea of having _some_ fallback, but it should at least be maximally conservative!

Note that `azcopy` can already be configured by [environment variables](https://docs.microsoft.com/en-us/azure/storage/common/storage-ref-azcopy-configuration-settings), including `AZCOPY_CONCURRENCY_VALUE`. It also does its own concurrency auto-tuning, capturing one of the proposals you sketched out.

WDYT?

**Comment by puhley** [comment]
Looking at the code, both` upload_file` and `upload_file_data` use `max_concurrency`. The functional difference between the two is that `upload_file` takes a file path and `upload_file_data` takes the data as a parameter.  I can't find anywhere where `upload_file_data` is actually used in the OneFuzz code. It appears to have been added for logical completeness in the class but it was never needed. The `max_concurrency` may not be as big of an issue with `upload_file_data` since it is likely that data being passed as a parameter would be smaller in size. Although, it may make sense to remove `max_concurrency` from both to keep the behavior consistent.

Based on your recommendations, my initial interpretation is that the new logic would be:

```
    @retry(
        stop=stop_after_attempt(10),
        wait=wait_random(min=1, max=3),
        retry=retry_if_exception_type(),
        before_sleep=before_sleep,
        reraise=True,
    )
    def upload_file(self, file_path: str, blob_name: str) -> None:
        try:
            # First try with azcopy if it is installed
            azcopy_sync(file_path, self.container_url + "/" + blob_name)
        except Exception as exc:
            # This will show either the OneFuzz warning that azcopy was not found or the subprocess error from the command
            LOGGER.warning(exc)
            # Indicate a change in approach so that future bug reports are clear about which of the two approaches is failing
            LOGGER.warning("Falling back to Python SDK...")
 
            # This does not have a try/except since it should be caught by the retry system.
            # The retry system would always trigger azcopy first and this approach second
            with open(file_path, "rb") as handle:
                # This has max_concurrency removed so that it uses the Azure SDK default
                self.client.upload_blob(
                     name=blob_name, data=handle, overwrite=True
                )
        return None
```


I haven't yet tested the above code. This is only a rough draft with extra commenting for discussion purposes. If this code aligns with your vision, then I can perform some local testing and create a pull request based on it.

**Comment by ranweiler** [comment]
@puhley, sorry for the delayed reply, somehow I missed your follow-up in notification noise!

What you sketched is consistent with what I had in mind.

I think it makes sense to remove `max_concurrency` from both `upload_file..` methods, and even to have them use the same logic of "prefer `azcopy`, fallback to SDK if needed".

**Comment by ranweiler** [comment]
Closed by #1556.
